### ISSUE TITLE
Symbol Map / Deminifier

### ISSUE THREAD

**Issue by malware-dev** [malforge/mdk2#32]
Include a map of the symbols and create a tool to deminify a script / create an MDK project from a script blueprint

- symbol map must be optional as some people use the minifiers as a base-level obfuscator.

**Comment by dosluke** [comment]
i like this but... tbh if the creator of the script wants they can post it in github.. if they dont want to, they arnt gonna share the symbol map either, most likely.

**Comment by malware-dev** [comment]
Correct, but the minifier _isn't_ an obfuscator tool, and a fair few of us actually want people to be able to learn from what we do.

**Comment by slammingprogramming** [comment]
I think adding a deminifier tool with an optional symbol map would be a valuable feature for the community, especially for those looking to learn from more complex scripts. While it's true that some creators may prefer not to share their symbol maps, making the feature optional keeps control in their hands while still supporting those who want to promote educational use and code transparency. For users who want to understand how advanced scripts work or troubleshoot them, having a way to “reverse engineer” an MDK project from a minified script could be a big step forward, especially given the heavy optimization and custom methods involved in many Space Engineers scripts.

It's also not entirely out of the question to add an obfuscator alongside the minifier, which could help protect more sensitive scripts while still allowing this deminifier feature for creators focused on transparency. Even then, I wonder—what really stops someone from deobfuscating any given script if they're determined? Obfuscation can make it harder to read, but it’s not a failsafe against reverse engineering.

Finally, for more complex scripts with nested classes or regions, any thoughts on how a deminifier tool might handle symbol maps in those cases? The structure could get tricky, but if done right, this could really streamline the process for creators who want to share both clear and optimized versions without constantly managing separate codebases.

**Comment by dosluke** [comment]
Yeah I was working on understanding a minified script with 0 source code otherwise, and it was going very slowly.. and then i tried pasting a whole class into chatgpt and told it to guess at what all the symbols were/meant, and i gave it some context and viola it actually got it majorly right as far as i could tell. made sense to me.

**Comment by slammingprogramming** [comment]
That’s actually a great point, dosluke! AI tools can be surprisingly effective at interpreting code when given enough context, even with minified scripts. It’s impressive that ChatGPT was able to make accurate guesses about symbol meanings—that highlights just how helpful an integrated deminifier with an optional symbol map could be for those looking to reverse-engineer scripts for learning purposes.

Incorporating a deminifier directly in MDK, especially one designed to recognize and preserve class and method structures, would provide a much more streamlined and accessible solution. Users wouldn’t need to rely on external tools or guesswork to understand symbol meanings, and they’d have a clearer, more reliable starting point when analyzing complex scripts.

Plus, with the deminifier designed specifically for MDK and Space Engineers scripts, it could account for quirks unique to this environment, like custom region directives or specific naming conventions. This could make the tool particularly valuable, even for users familiar with AI assistance, by delivering more tailored and contextually accurate results.

Ultimately, having both an optional symbol map and the choice to obfuscate would allow creators to balance transparency with control over their code’s accessibility. If nothing else, the feedback we’re discussing here suggests there’s a lot of interest in tools that make advanced scripting easier to understand and share.
